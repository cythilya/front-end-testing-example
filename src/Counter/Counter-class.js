import { Component } from 'react';

class Counter extends Component {
  state = { count: 0 };

  increment = () => {
    this.setState((prevState) => ({
      count: prevState.count + 1,
    }));
  };

  decrement = () => {
    this.setState((prevState) => ({
      count: prevState.count - 1,
    }));
  };

  render() {
    return (
      <>
        <p data-test-id="counter-value">Count: {this.state.count}</p>
        <button data-test-id="decrement-button" onClick={this.decrement}>
          +
        </button>
        <button data-test-id="increment-button" onClick={this.increment}>
          -
        </button>
      </>
    );
  }
}

export default Counter;
