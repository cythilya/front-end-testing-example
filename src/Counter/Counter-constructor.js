class Counter {
  constructor() {
    this.count = 0;
  }

  increment() {
    this.count++;
  }

  decrement() {
    this.count--;
  }

  getCount() {
    return this.count;
  }
}

module.exports = Counter;
