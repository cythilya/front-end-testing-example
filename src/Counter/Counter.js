import { useState } from 'react';

const Counter = () => {
  const [count, setCount] = useState(0);
  const increment = () => setCount(count + 1);
  const decrement = () => setCount(count - 1);

  return (
    <>
      <div data-test-id="counter-value">{count}</div>
      <button data-test-id="decrement-button" onClick={decrement}>
        -
      </button>
      <button data-test-id="increment-button" onClick={increment}>
        +
      </button>
    </>
  );
};

export default Counter;
