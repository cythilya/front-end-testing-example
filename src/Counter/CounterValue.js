const CounterValue = ({ count }) => <p data-test-id="counter-value">{count}</p>;

export default CounterValue;