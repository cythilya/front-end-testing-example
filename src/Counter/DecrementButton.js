const DecrementButton = ({ onClick }) => {
  return (
    <button data-test-id="decrement-button" onClick={onClick}>
      -
    </button>
  );
};

export default DecrementButton;