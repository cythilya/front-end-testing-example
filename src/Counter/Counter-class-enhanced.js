import { Component } from 'react';
import CounterValue from './CounterValue';
import DecrementButton from './DecrementButton';
import IncrementButton from './IncrementButton';

class Counter extends Component {
  state = { count: 0 };

  increment = () => {
    this.setState((prevState) => ({
      count: prevState.count + 1,
    }));
  };

  decrement = () => {
    this.setState((prevState) => ({
      count: prevState.count - 1,
    }));
  };

  render() {
    return (
      <>
        <CounterValue count={this.state.count} />
        <DecrementButton onClick={this.decrement} />
        <IncrementButton onClick={this.increment} />
      </>
    );
  }
}

export default Counter;
