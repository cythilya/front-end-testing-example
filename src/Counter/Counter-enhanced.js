import { useState } from 'react';
import CounterValue from './CounterValue';
import DecrementButton from './DecrementButton';
import IncrementButton from './IncrementButton';

const Counter = () => {
  const [count, setCount] = useState(0);
  const increment = () => setCount(count + 1);
  const decrement = () => setCount(count - 1);

  return (
    <>
      <CounterValue count={count} />
      <DecrementButton onClick={decrement} />
      <IncrementButton onClick={increment} />
    </>
  );
};

export default Counter;
