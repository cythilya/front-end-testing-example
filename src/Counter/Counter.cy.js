import React from 'react'
import Counter from './Counter'

describe('<Counter />', () => {
  it('should get 1 when click the increment button', () => {
    cy.mount(<Counter />)

    cy.get('[data-test-id="increment-button"]').click();

    cy.get('[data-test-id="counter-value"]').should('have.text', '1');
  })

  it('should get -1 when click the decrement button', () => {
    cy.mount(<Counter />)

    cy.get('[data-test-id="decrement-button"]').click();

    cy.get('[data-test-id="counter-value"]').should('have.text', '-1');
  });
})
