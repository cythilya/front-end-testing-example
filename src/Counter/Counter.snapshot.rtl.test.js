import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { createRenderer } from 'react-test-renderer/shallow';
import Counter from './Counter-enhanced';

describe('Counter component', () => {
  describe('shallowing rendering', () => {
    let counter;
    const renderer = createRenderer();
    const renderCounter = () => renderer.render(<Counter />);

    beforeEach(() => {
      counter = renderCounter();
    });

    afterEach(() => {
      renderer.unmount();
    });

    it('should get snapshot correctly', () => {
      const counter = renderer.render(<Counter />);

      expect(counter).toMatchSnapshot();
    });
  });

  describe('full rendering', () => {
    let counter;
    const renderCounter = () => <Counter />;

    beforeEach(() => {
      counter = render(renderCounter());
    });

    afterEach(() => {
      counter.unmount();
    });

    it('should get snapshot correctly', () => {
      expect(counter.container).toMatchSnapshot();
    });

    it('should get 1 when click the increment button', () => {
      const { getByTestId } = counter;

      fireEvent.click(getByTestId('increment-button'));

      expect(getByTestId('counter-value')).toHaveTextContent('1');
    });
  });
});
