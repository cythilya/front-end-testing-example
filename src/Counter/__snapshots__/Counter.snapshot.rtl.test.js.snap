// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component full rendering should get snapshot correctly 1`] = `
<div>
  <p
    data-test-id="counter-value"
  >
    0
  </p>
  <button
    data-test-id="decrement-button"
  >
    -
  </button>
  <button
    data-test-id="increment-button"
  >
    +
  </button>
</div>
`;

exports[`Counter component shallowing rendering should get snapshot correctly 1`] = `
<React.Fragment>
  <CounterValue
    count={0}
  />
  <DecrementButton
    onClick={[Function]}
  />
  <IncrementButton
    onClick={[Function]}
  />
</React.Fragment>
`;
